File: android/src/main/java/com/getcapacitor/community/firebaseanalytics/FirebaseAnalytics.kt
package com.getcapacitor.community.firebaseanalytics

import android.Manifest
import android.os.Bundle
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginException
import com.getcapacitor.PluginMethod
import com.getcapacitor.PluginThread
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.google.firebase.analytics.FirebaseAnalytics as GoogleFirebaseAnalytics
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

@CapacitorPlugin(
    name = "FirebaseAnalytics",
    permissions = [
        Permission(strings = [Manifest.permission.ACCESS_NETWORK_STATE], alias = "network"),
        Permission(strings = [Manifest.permission.INTERNET], alias = "internet"),
        Permission(strings = [Manifest.permission.WAKE_LOCK], alias = "wakelock")
    ]
)
public class FirebaseAnalytics : Plugin() {
    private var firebaseAnalytics: GoogleFirebaseAnalytics? = null

    override fun load() {
        super.load()

        // Obtain the FirebaseAnalytics instance.
        firebaseAnalytics = GoogleFirebaseAnalytics.getInstance(bridge.activity)
    }

    /**
     * Sets the user ID property.
     * @param call - userId: unique identifier of the user to log
     */
    @PluginMethod
    public fun setUserId(call: PluginCall) {
        val analytics = firebaseAnalytics ?: throw PluginException(MISSING_REF_MSSG)

        if (!call.data.has("userId")) {
            throw PluginException("userId property is missing")
        }

        try {
            analytics.setUserId(call.getString("userId"))
            call.resolve()
        } catch (ex: Exception) {
            call.reject(ex.localizedMessage)
        }
    }

    /**
     * Sets a user property to a given value.
     * @param call - name: The name of the user property to set.
     *               value: The value of the user property.
     */
    @PluginMethod
    public fun setUserProperty(call: PluginCall) {
        val analytics = firebaseAnalytics ?: throw PluginException(MISSING_REF_MSSG)

        if (!call.data.has("name")) {
            throw PluginException("name property is missing")
        }

        if (!call.data.has("value")) {
            throw PluginException("value property is missing")
        }

        // Firebase requires a name. One that is present but not a string is reported like a missing one.
        val name = call.getString("name") ?: throw PluginException("name property is missing")

        try {
            val value = call.getString("value")

            analytics.setUserProperty(name, value)
            call.resolve()
        } catch (ex: Exception) {
            call.reject(ex.localizedMessage)
        }
    }

    /**
     * Retrieves the app instance id from the service.
     * @param call - instanceId: current instance if of the app
     */
    @PluginMethod
    public fun getAppInstanceId(call: PluginCall) {
        val analytics = firebaseAnalytics ?: throw PluginException(MISSING_REF_MSSG)
        analytics.appInstanceId.addOnCompleteListener { task ->
            if (task.isSuccessful) {
                val instanceId = task.result
                if (instanceId != null && instanceId.isEmpty()) {
                    call.reject("failed to obtain app instance id")
                } else {
                    val result = JSObject()
                    result.put("instanceId", instanceId)
                    call.resolve(result)
                }
            } else {
                call.reject(task.exception?.localizedMessage)
            }
        }
    }

    /**
     * Sets the current screen name, which specifies the current visual context in your app.
     * @param call - screenName: the activity to which the screen name and class name apply.
     *               nameOverride: the name of the current screen. Set to null to clear the current screen name.
     */
    // Logs the screen_view event on the main thread, as it did from a runOnUiThread block.
    @PluginMethod(thread = PluginThread.MAIN)
    public fun setScreenName(call: PluginCall) {
        val analytics = firebaseAnalytics ?: throw PluginException(MISSING_REF_MSSG)

        if (!call.data.has("screenName")) {
            throw PluginException("screenName property is missing")
        }

        try {
            val screenName = call.getString("screenName")
            val nameOverride = call.getString("nameOverride", null)

            val bundle = Bundle()
            bundle.putString(GoogleFirebaseAnalytics.Param.SCREEN_NAME, screenName)
            bundle.putString(GoogleFirebaseAnalytics.Param.SCREEN_CLASS, nameOverride)
            analytics.logEvent(GoogleFirebaseAnalytics.Event.SCREEN_VIEW, bundle)
            call.resolve()
        } catch (ex: Exception) {
            call.reject(ex.localizedMessage)
        }
    }

    /**
     * Clears all analytics data for this app from the device and resets the app instance id.
     * @param call
     */
    @PluginMethod
    public fun reset(call: PluginCall) {
        val analytics = firebaseAnalytics ?: throw PluginException(MISSING_REF_MSSG)

        try {
            analytics.resetAnalyticsData()
            call.resolve()
        } catch (ex: Exception) {
            call.reject(ex.localizedMessage)
        }
    }

    /**
     * Logs an app event.
     * @param call - name: unique name of the event
     *               params: the map of event parameters.
     */
    @PluginMethod
    public fun logEvent(call: PluginCall) {
        val analytics = firebaseAnalytics ?: throw PluginException(MISSING_REF_MSSG)

        if (!call.data.has("name")) {
            throw PluginException("name property is missing")
        }

        // Firebase requires a name. One that is present but not a string is reported like a missing one.
        val name = call.getString("name") ?: throw PluginException("name property is missing")

        try {
            val params = call.data.getJSObject("params")
            analytics.logEvent(name, params?.let { convertJsonToBundle(it) })
            call.resolve()
        } catch (ex: Exception) {
            call.reject(ex.localizedMessage)
        }
    }

    /**
     * Sets whether analytics collection is enabled for this app on this device.
     * @param call - enabled: boolean true/false to enable/disable logging
     */
    @PluginMethod
    public fun setCollectionEnabled(call: PluginCall) {
        val analytics = firebaseAnalytics ?: throw PluginException(MISSING_REF_MSSG)

        val enabled = call.getBoolean("enabled", false) ?: false

        analytics.setAnalyticsCollectionEnabled(enabled)
        call.resolve()
    }

    /**
     * Deprecated: use setCollectionEnabled() instead
     * Enable analytics collection for this app on this device.
     * @param call - enabled: boolean true/false to enable/disable logging
     */
    @Deprecated("Use setCollectionEnabled() instead")
    @PluginMethod
    public fun enable(call: PluginCall) {
        val analytics = firebaseAnalytics ?: throw PluginException(MISSING_REF_MSSG)

        analytics.setAnalyticsCollectionEnabled(true)
        call.resolve()
    }

    /**
     * Deprecated: use setCollectionEnabled() instead
     * Disable analytics collection for this app on this device.
     * @param call
     */
    @Deprecated("Use setCollectionEnabled() instead")
    @PluginMethod
    public fun disable(call: PluginCall) {
        val analytics = firebaseAnalytics ?: throw PluginException(MISSING_REF_MSSG)

        analytics.setAnalyticsCollectionEnabled(false)
        call.resolve()
    }

    /**
     * Sets the duration of inactivity that terminates the current session.
     * @param call: options - duration: duration of inactivity
     */
    @PluginMethod
    public fun setSessionTimeoutDuration(call: PluginCall) {
        val analytics = firebaseAnalytics ?: throw PluginException(MISSING_REF_MSSG)

        val duration = call.getInt("duration", 1800) ?: 1800

        analytics.setSessionTimeoutDuration(duration.toLong())
        call.resolve()
    }

    public companion object {
        private const val MISSING_REF_MSSG = "Firebase analytics is not initialized"

        @JvmStatic
        public fun convertJsonToBundle(json: JSONObject?): Bundle {
            val bundle = Bundle()
            if (json == null || json.length() == 0) return bundle

            for (key in json.keys()) {
                try {
                    when (val value = json.get(key)) {
                        is String -> bundle.putString(key, value)
                        is Boolean -> bundle.putBoolean(key, value)
                        is Int -> bundle.putInt(key, value)
                        is Long -> bundle.putLong(key, value)
                        is Float -> bundle.putFloat(key, value)
                        is Double -> bundle.putDouble(key, value)
                        is JSONObject -> bundle.putBundle(key, convertJsonToBundle(value))
                        is JSONArray -> putArray(bundle, key, value)
                    }
                } catch (e: ClassCastException) {
                    e.printStackTrace()
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            }

            return bundle
        }

        // The first element decides the type of the array; an element that does not fit throws, and the
        // caller then leaves the key out.
        @Throws(JSONException::class)
        private fun putArray(bundle: Bundle, key: String, array: JSONArray) {
            when (if (array.length() == 0) null else array.get(0)) {
                is JSONObject -> bundle.putParcelableArray(key, Array(array.length()) { convertJsonToBundle(array.getJSONObject(it)) })

                is String -> bundle.putStringArray(key, Array(array.length()) { array.getString(it) })

                is Int, is Float, is Double ->
                    bundle.putFloatArray(key, FloatArray(array.length()) { (array.get(it) as Number).toFloat() })
            }
        }
    }
}

File: ios/Sources/FirebaseAnalyticsPlugin/FirebaseAnalytics.swift
import Foundation
import Capacitor
import FirebaseCore
import FirebaseAnalytics

@objc(FirebaseAnalytics)
public class FirebaseAnalytics: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "FirebaseAnalytics"
    public let jsName = "FirebaseAnalytics"
    public let pluginMethods: [CAPPluginMethod] = [
        .promise("setUserId", FirebaseAnalytics.setUserId),
        .promise("setUserProperty", FirebaseAnalytics.setUserProperty),
        .promise("getAppInstanceId", FirebaseAnalytics.getAppInstanceId),
        .promise("setScreenName", FirebaseAnalytics.setScreenName),
        .promise("reset", FirebaseAnalytics.reset),
        .promise("logEvent", FirebaseAnalytics.logEvent),
        .promise("setCollectionEnabled", FirebaseAnalytics.setCollectionEnabled),
        .promise("setSessionTimeoutDuration", FirebaseAnalytics.setSessionTimeoutDuration),
        .promise("enable", FirebaseAnalytics.enable),
        .promise("disable", FirebaseAnalytics.disable)
    ]

    // Every method stays synchronous: the bridge queue runs the setters and the logged events in the order of the
    // calls, which async methods would not keep.

    public override func load() {
        if FirebaseApp.app() == nil {
            FirebaseApp.configure()
        }
    }

    /// Sets the user ID property.
    /// - Parameter call: userId - unique identifier of the user to log
    func setUserId(_ call: CAPPluginCall) throws {
        guard let userId = call.getString("userId") else {
            throw CAPPluginError("userId property is missing")
        }
        Analytics.setUserID(userId)
        call.resolve()
    }

    /// Sets a user property to a given value.
    /// - Parameter call: name - The name of the user property to set.
    ///                   value - The value of the user property.
    func setUserProperty(_ call: CAPPluginCall) throws {
        guard let name = call.getString("name") else {
            throw CAPPluginError("name property is missing")
        }
        guard let value = call.getString("value") else {
            throw CAPPluginError("value property is missing")
        }
        Analytics.setUserProperty(value, forName: name)
        call.resolve()
    }

    /// Retrieves the app instance id from the service.
    /// - Parameter call: instanceId - current instance if of the app
    func getAppInstanceId(_ call: CAPPluginCall) {
        let instanceId = Analytics.appInstanceID()
        call.resolve([
            "instanceId": instanceId
        ])
    }

    /// Sets the current screen name, which specifies the current visual context in your app.
    /// - Parameter call: screenName - the activity to which the screen name and class name apply.
    ///                   nameOverride - the name of the current screen. Set to null to clear the current screen name.
    func setScreenName(_ call: CAPPluginCall) throws {
        guard let screenName = call.getString("screenName") else {
            throw CAPPluginError("screenName property is missing")
        }
        let nameOverride = call.getString("nameOverride") ?? nil
        DispatchQueue.main.async {
            Analytics.logEvent(AnalyticsEventScreenView,
                parameters: [AnalyticsParameterScreenName: screenName,
                            AnalyticsParameterScreenClass: nameOverride])
        }
        call.resolve()
    }


    /// Clears all analytics data for this app from the device and resets the app instance id.
    func reset(_ call: CAPPluginCall) {
        Analytics.resetAnalyticsData()
        call.resolve()
    }


    /// Logs an app event.
    /// - Parameter call: name - unique name of the event
    ///                   params - the map of event parameters.
    func logEvent(_ call: CAPPluginCall) throws {

        /// Name is a required argument to logEvent()
        guard let name = call.getString("name"), !name.isEmpty else {
            throw CAPPluginError("Event name is required and can't be empty")
        }

        /// logEvent() expects `nil` when there are no parameters
        guard var params = call.getObject("params"), !params.isEmpty else {
            Analytics.logEvent(name, parameters: nil)
            call.resolve()
            return
        }

        /// FirebaseAnalytics silently converts any item quantity that is not an
        /// integer to zero, this includes any NSNumber or string value passed
        /// as an option to CAPPluginCall.
        if var items = params["items"] as? NSArray as? [[String:Any]] {
            for (idx, item) in items.enumerated() {
                if let quantity = item["quantity"] {
                    guard let intVal = quantity as? Int else {
                        throw CAPPluginError("Item quantity must be specified as an integer value")
                    }
                    items[idx]["quantity"] = intVal
                }
            }
            params["items"] = items
        }

        Analytics.logEvent(name, parameters: params)
        call.resolve()
    }


    /// Sets whether analytics collection is enabled for this app on this device.
    /// - Parameter call: enabled - boolean true/false to enable/disable logging
    func setCollectionEnabled(_ call: CAPPluginCall) {
        if let enabled = call.getBool("enabled") {
            Analytics.setAnalyticsCollectionEnabled(enabled)
        } else {
            Analytics.setAnalyticsCollectionEnabled(false)
        }
        call.resolve()
    }


    /// Sets the duration of inactivity that terminates the current session.
    /// - Parameter call: duration - duration of inactivity
    func setSessionTimeoutDuration(_ call: CAPPluginCall) {
        let duration = call.getInt("duration") ?? 1800

        Analytics.setSessionTimeoutInterval(TimeInterval(duration))
        call.resolve()
    }

    /// Deprecated - use setCollectionEnabled instead
    /// Enable analytics collection for this app on this device.
    /// - Parameter call
    func enable(_ call: CAPPluginCall) {
        Analytics.setAnalyticsCollectionEnabled(true)
        call.resolve()
    }

    /// Deprecated - use setCollectionEnabled instead
    /// Disable analytics collection for this app on this device.
    /// - Parameter call
    func disable(_ call: CAPPluginCall) {
        Analytics.setAnalyticsCollectionEnabled(false)
        call.resolve()
    }
}
